File: src/javascript/algorithms/hashtable/firstNonRepeatingCharacter.js
function firstNonRepeatingCharWithMap(inputStr) {
    if(inputStr.length === 0) return null;
    let charCountsMap = new Map();
    for(let ch of inputStr) {
        charCountsMap.set(ch, (charCountsMap.get(ch) || 0) + 1);
    }
    
    for(let [key, value] of charCountsMap.entries()) {
        if(value === 1) {
            return key;
        }
    }
    return null;
}

function firstNonRepeatingCharWithObj(inputStr) {
    if(inputStr.length === 0) return null;
    let charCountsObj = {};
    for(let ch of inputStr) {
        charCountsObj[ch] = (charCountsObj[ch] || 0) + 1;
    }
    
    for(let [key, value] of Object.entries(charCountsObj)) {
        if(value === 1) {
            return key;
        }
    }
    return null;
}



/**---------------
   All Unique
   --------------- */
console.log("Input: 'abcdefgh'");
console.log("Output: ", firstNonRepeatingCharWithObj('abcdefgh'));
console.log("---------------");

/**---------------
   Some Duplicates
   --------------- */
console.log("Input: 'abccddefgggh'");
console.log("Output: ", firstNonRepeatingCharWithObj('abccddefgggh'));
console.log("---------------");

/** ---------------
    All Duplicates
    --------------- */
console.log("Input: 'aabbccdddeeeff'");
console.log("Output: ", firstNonRepeatingCharWithMap('aabbccdddeeeff'));
console.log("---------------");

/** ---------------
    Empty String
    --------------- */
console.log("Input: ''");
console.log("Output: ", firstNonRepeatingMap(''));
console.log("---------------");

File: src/java1/algorithms/hashmap/firstNonRepeatingCharacter/FirstNonRepeatingCharacter.java
package java1.algorithms.hashmap.firstNonRepeatingCharacter;

import java.util.HashMap;
import java.util.Map;

public class FirstNonRepeatingCharacter {
    private static Character firstNonRepeatingChar(String str){
        if(str.length() == 0) return null;

        Map<Character, Integer> charCountsMap = new HashMap<>();
        for(char ch: str.toCharArray()){
            charCountsMap.put(ch, charCountsMap.getOrDefault(ch, 0) +1 );
        }

        for(Map.Entry<Character, Integer> entry: charCountsMap.entrySet()){
            if(entry.getValue() == 1) {
                return entry.getKey();
            }
        }

        return null;
    }
    public static void main(String[] args) {
        /**---------------
        All Unique
        --------------- */
        System.out.println("Input: 'abcdefgh'");
        System.out.println("Output: "+ firstNonRepeatingChar("abcdefgh"));
        System.out.println("---------------");

        /**---------------
         Some Duplicates
        --------------- */
        System.out.println("Input: 'abccddefgggh'");
        System.out.println("Output: "+ firstNonRepeatingChar("abccddefgggh"));
        System.out.println("---------------");

        /** ---------------
            All Duplicates
            --------------- */
        System.out.println("Input: 'aabbccdddeeeff'");
        System.out.println("Output: "+ firstNonRepeatingChar("aabbccdddeeeff"));
        System.out.println("---------------");

        /** ---------------
            Empty String
            --------------- */
        System.out.println("Input: ''");
        System.out.println("Output: "+ firstNonRepeatingChar(""));
        System.out.println("---------------");
    }
}
